numpyro/distributions/distribution.py

```diff
@@ -774,16 +762,10 @@ def __init__(self, base_distribution, transforms, validate_args=None):
             raise ValueError("Base distribution needs to have shape with size at least {}, but got {}."
                              .format(transform_input_event_dim, base_ndim))
         event_dim = transform.codomain.event_dim + max(self.base_dist.event_dim - transform_input_event_dim, 0)
-        # See the above note. Currently, there is no way to interpret the shape of output after
-        # transforming. To solve this issue, we need something like Bijector.forward_event_shape
-        # as in TFP. For now, we will prepend singleton dimensions to compromise, so that
-        # event_dim, len(batch_shape) are still correct.
-        if event_dim <= base_ndim:
-            batch_shape = shape[:base_ndim - event_dim]
-            event_shape = shape[base_ndim - event_dim:]
-        else:
-            event_shape = (-1,) * event_dim
-            batch_shape = ()
+        shape = transform.forward_shape(shape)
+        cut = len(shape) - event_dim
+        batch_shape = shape[:cut]
+        event_shape = shape[cut:]
```

Comment:
Nice, this is so much simpler that the PyTorch version! I think I'll copy your use of `ComposeTransform` upstream in https://github.com/pytorch/pytorch/pull/50581 😄

Reply:
@fehiepsi I've made some more changes to the upstream version, including wrapping `base_dist` in `Independent` to ensure `.sample(sample_shape)` is correct. Can you PTAL at https://github.com/pytorch/pytorch/pull/50581?

Reply:
You are right. We have assumed too much on base distribution. Your change is pretty clean. Thanks!